ipfn/ipfn.py

```diff
@@ -153,83 +152,73 @@ def ipfn_df(self, df, aggregates, dimensions, weight_col='total'):
-            table_current.set_index(features, inplace=True)
-
-            for feature in product(*feat_l):
-                den = tmp.loc[feature]
-                # calculate new weight for this iteration
-                if den == 0:
-                    table_update.loc[feature, weight_col] =\
-                        table_current.loc[feature, weight_col] *\
-                        xijk.loc[feature]
-                else:
-                    table_update.loc[feature, weight_col] = \
-                        table_current.loc[feature, weight_col].astype(float) * \
-                        xijk.loc[feature]/den
-
-            table_update.reset_index(inplace=True)
-            table_current.reset_index(inplace=True)
-            inc += 1
-            feat_l = []
+            # Store the convergence factors
+            f = f.sub(1).abs().max()
+            factors.append(f)
 
-        # Calculate the max convergence rate
-        max_conv = 0
-        inc = 0
-        for features in dimensions:
-            tmp = df.groupby(features)[weight_col].sum()
-            ori_ijk = aggregates[inc]
-            temp_conv = max(abs(tmp/ori_ijk - 1))
-            if temp_conv > max_conv:
-                max_conv = temp_conv
-            inc += 1
+        # Check for convergence
+        max_conv = max(factors)
 
         return df, max_conv
 
     def iteration(self):
         """
-        Runs the ipfn algorithm. Automatically detects of working with numpy ndarray or pandas dataframes.
+        Runs the ipfn algorithm. Automatically detects of working with
+        numpy ndarray or pandas dataframes.
         """
 
         i = 0
         conv = self.conv_rate * 100
-        m = self.original
+        m = self.original.copy()
 
         # If the original data input is in pandas DataFrame format
         if isinstance(self.original, pd.DataFrame):
+            # Add index
+            indexcols = list(set(x for l in self.dimensions for x in l))
+            m.set_index(indexcols, inplace=True)
+            # Turn to series
+            m = m[self.ipfn_column]
             while i <= self.max_itr and conv > self.conv_rate:
-                m, conv = self.ipfn_df(m, self.aggregates, self.dimensions, self.weight_col)
+                m, conv = self.ipfn_df(m, self.aggregates, self.dimensions)
                 i += 1
                 # print(i, conv)
         # If the original data input is in numpy format
         elif isinstance(self.original, np.ndarray):
             self.original = self.original.astype('float64')
             while i <= self.max_itr and conv > self.conv_rate:
-                m, conv = self.ipfn_np(m, self.aggregates, self.dimensions, self.weight_col)
+                m, conv = self.ipfn_np(m, self.aggregates, self.dimensions)
                 i += 1
                 # print(i, conv)
+
+        # Reset to dataframe
+        m.name = self.ipfn_column
+        df = m.reset_index()
```

Comment:
this fails is the ipf is performed on numpy arrays as opposed to pandas dataframes.
The 2 lines 219 and 220 should go under the **if isinstance(self.original, pd.DataFrame)** condition and line 220 should become
**m = m.reset_index()**